csa.rs

```diff
@@ -0,0 +1,134 @@
+            in_connection[connection.arrival_station] = i as usize;
+            
+            if connection.arrival_station == arrival_station && connection.arrival_timestamp < earliest {
+                earliest = connection.arrival_timestamp;
+            }
+        } else if connection.arrival_timestamp > earliest {
+            return;                    
+        }                    
+    }
+}
+
+fn csa_print_result(timetable: &Vec<Connection>, in_connection: &[usize; MAX_STATIONS], arrival_station: usize) {
+    if in_connection[arrival_station] == std::u32::MAX as usize {
+        println!("NO_SOLUTION");
+    } else {
+        let mut route = Vec::new();
+        let mut last_connection_index = in_connection[arrival_station];
+        
+        while last_connection_index != std::u32::MAX as usize {
+            let ref connection = timetable[last_connection_index];
+            route.push(connection);
+            last_connection_index = in_connection[connection.departure_station];
+        }
+        
+        for connection in route.iter().rev() {
+            println!("{} {} {} {}", connection.departure_station, connection.arrival_station, connection.departure_timestamp, connection.arrival_timestamp); 
+        }
+    }           
+    println!("");
+}
+
+fn csa_compute(timetable: &Vec<Connection>, departure_station: usize, arrival_station: usize, departure_time: u32)
+{ 
+    let mut in_connection = [std::u32::MAX as usize; MAX_STATIONS];
+    let mut earliest_arrival = [std::u32::MAX; MAX_STATIONS];
+            
+    earliest_arrival[departure_station as usize] = departure_time;
+    
+    if departure_station <= MAX_STATIONS && arrival_station <= MAX_STATIONS {
+        csa_main_loop(&timetable, arrival_station, &mut earliest_arrival, &mut in_connection);
+    }
+    
+    csa_print_result(&timetable, &in_connection, arrival_station);
+}
+
+fn main() {
+    let mut timetable = Vec::<Connection>::new();
+    
+    // Importing connections
+    
+    loop {
+        let mut input_line = String::new();
+        io::stdin().read_line(&mut input_line)
+            .ok()
+            .expect("failed to read connection line");
+            
+        input_line = input_line.trim_right().to_owned(); 
+            
+        if input_line.is_empty() {
+            break; 
+        } else {
+            timetable.push(Connection::parse(input_line)); 
+        };
+    }  
+    
+    // Responding to requests from stdin
+    
+    loop {
+        let mut input_line = String::new();
+        io::stdin().read_line(&mut input_line)
+            .ok()
+            .expect("failed to read request line");
+            
+        input_line = input_line
+            .trim_right()
+            .to_owned();
+            
+        if input_line.is_empty() {
+            break;        
+        } else {        
+            let params = input_line.split(" ")
+                .map(|bite| {
+                    bite.parse::<u32>()
```

Comment:
explicit type instancing should be optional here as it is provided to `collect()`